flair/embeddings.py

```diff
@@ -2213,199 +2215,113 @@ def __init__(
-            )
-        )
+        longest_sequence_in_batch: int = len(max(tokenized, key=len))
 
-        # prepare id maps for BERT model
-        features = self._convert_sentences_to_features(
-            sentences, longest_sentence_in_batch
-        )
-        all_input_ids = torch.LongTensor([f.input_ids for f in features]).to(
-            flair.device
+        # initialize batch tensors
+        input_ids = torch.zeros(
+            [len(sentences), longest_sequence_in_batch],
+            dtype=torch.long,
+            device=flair.device,
         )
-        all_input_masks = torch.LongTensor([f.input_mask for f in features]).to(
-            flair.device
+        mask = torch.zeros(
+            [len(sentences), longest_sequence_in_batch],
+            dtype=torch.long,
+            device=flair.device,
         )
 
-        # put encoded batch through BERT model to get all hidden states of all encoder layers
-        self.model.to(flair.device)
-        self.model.eval()
-        all_encoder_layers = self.model(all_input_ids, attention_mask=all_input_masks)[
-            -1
-        ]
+        for s_id, sentence in enumerate(tokenized):
+            sequence_length = len(sentence)
+            input_ids[s_id][:sequence_length] = sentence
+            mask[s_id][:sequence_length] = torch.ones(sequence_length)
+
 
-        with torch.no_grad():
+        # put encoded batch through BERT model to get all hidden states of all encoder layers
+        all_encoder_layers = self.model(input_ids, attention_mask=mask)[-1]
 
-            for sentence_index, sentence in enumerate(sentences):
+        # gradients are enable if fine-tuning is enabled
+        gradient_context = torch.enable_grad() if (self.fine_tune and self.training) else torch.no_grad()
 
-                feature = features[sentence_index]
+        with gradient_context:
+            for sentence_idx, (subtoken_ids, sentence) in enumerate(zip(tokenized, sentences)):
+                subtokens = self.tokenizer.convert_ids_to_tokens(subtoken_ids)
+                
+                token_subtoken_embeddings = []
+                token_idx = 0
+                for subtoken_idx, subtoken in enumerate(subtokens):
+                    if subtoken == '[SEP]':
+                        continue
 
-                # get aggregated embeddings for each BERT-subtoken in sentence
-                subtoken_embeddings = []
-                for token_index, _ in enumerate(feature.tokens):
-                    all_layers = []
-                    for layer_index in self.layer_indexes:
-                        if self.use_scalar_mix:
-                            layer_output = all_encoder_layers[int(layer_index)][
-                                sentence_index
-                            ]
+                    # check if this token initiates a new word, internal subtokens start with ##
+                    if not subtoken.startswith('##') and len(token_subtoken_embeddings) > 0:
```

Comment:
This does not always work. For instance the code 

```python
train_sentence = Sentence('JAKARTATown 1996-08-27')
embeddings = BertEmbeddings(pooling_operation='first')
embeddings.embed(train_sentence)
```

Will break because Flair would consider this to be two tokens ("`JAKARTATown`" and "`1996-08-27`"), while BERT subtokenizes this as `['[CLS]', 'jakarta', '##town', '1996', '-', '08', '-', '27', '[SEP]']`. So `1996-08-27` is split but there are no `##` prefixes to indicate this.

@stefan-it you use the method `_get_transformer_sentence_embeddings` to go from subtokens to tokens for all transformer embeddings except for BERT. Can this method also be used for BERT or is there something different here?

Reply:
@alanakbik yes, it can be used here. The method should be robust enough to handle cases, where the BERT tokenizer discards tokens (e.g special characters, in this case the `<unk>` token is used to get a correct embedding for the "real" Flair token).

Reply:
Ah cool - I'll refactor it in, then! Do you have an example at hand for a case in which BERT tokenizer discards tokens?

Reply:
At the moment I've only seen `'\x96', '\u200e', '\x95', '\xad' or '\x80'` as problematic tokens (or better control characters:

```python
from transformers import BertTokenizer
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

tokenizer.tokenize("\x96")

# returns []
```

I found these tokens in the GermEval 2014 dataset (yes, this is bad) when fine-tuning BERT models with the Transformers NER code.

Reply:
I think we can heavily refactor the Transformer-base model inputs (like input ids, masks), because they're almost the same across all architectures. A good reference is the Hugging Face fine-tuning code for sequence labeling:

https://github.com/huggingface/transformers/blob/master/examples/ner/run_ner.py#L172-L178

Crucial part is always the different ways of tokenization - GPT-2 tokenization is horrible...

Reply:
Yes, I was just looking at the HuggingFace AutoModel classes: https://huggingface.co/transformers/model_doc/auto.html - based on this I'm trying to write a singular `TransformerWordEmbeddings` class for all transformer embeddings, instantiated with the model name. However, as you write, the big problem is matching all the different subtokenization schemes and their special symbols to our own tokens.